### PR TITLE
Andrewbladon/fix ogc feature service crash

### PR DESCRIPTION
**Pull request by AndrewBladon**
This PR is to fix crashes found during verification. Investigation into the cause suggested that if the load buttons were pressed while the feature service or a feature layer was loading, the sample viewer crashed. To prevent this, the load buttons have been disabled while the feature service or a feature layer is loading.

@ldanzinger and @JamesMBallard, could you confirm that, once approved, it is ok to merge these changes into the 100.12.0 release branch?

### CODE REVIEW

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml:133]: there is also `Enums.LoadStatusFailedToLoad` which you should handle as a terminal state

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml
Comment (lsmallwood) [inline ArcGISRuntimeSDKQt_QMLSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml:33]: I wonder if this could just be `property bool loading: featureService ? featureService.loadStatus === Enums.LoadStatusLoading : false`

**Review by lsmallwood — APPROVED** [review]
LGTM @AndrewBladon - a couple of thoughts for qml

File: ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.cpp
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.cpp:104]: our coding standard doesn't require brackets for one line bodys

```suggestion
  if (m_featureService && m_featureService->loadStatus() == LoadStatus::Loading)
    return;
```

File: ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.cpp
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.cpp:219]: It seems like there should be a simpler way where you don't need to be managing state throughout the code and setting/unsetting the loading state. Under which conditions would 'isLoading' be true? Is it just `if (m_featureService && m_featureService->loadStatus() == LoadStatus::Loading)`? Ideally we can let the API tell us when it is loading or not, but not sure if you have additional app level loading required here...

File: ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.h
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.h:91]: members should start with `m_` prefix

File: ArcGISRuntimeSDKQt_QMLSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml
Comment (AndrewBladon) [reply ArcGISRuntimeSDKQt_QMLSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml:133]: There is a function called `checkForLoadingErrors` which handles the `Enums.LoadStatusFailedToLoad`:
```
function checkForLoadingErrors() {
    if (root.featureService.loadStatus === Enums.LoadStatusFailedToLoad) {
        handleError(root.featureService.loadError);
    }
}
```
Will this be sufficient?

File: ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml:68]: if you are binding to an expression that returns bool, then it is redundant to use the ternary operator:

```suggestion
                    enabled: !model.loading
```

File: ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.qml:82]:
```suggestion
                    enabled: !model.loading
```

**Review by ldanzinger — APPROVED** [review]
I think there should be a cleaner way to do this than to manage the bool state throughout the app. Luke had a similar suggestion as to what I was thinking. I'm ok with this going in to fix the crash, but it should be reworked to streamline the workflow afterwards

**Comment by JamesMBallard** [comment]
>could you confirm that, once approved, it is ok to merge these changes into the 100.12.0 release branch?

Yes, we should fix a crash for the release. Good to merge once approved.

File: ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.cpp
Comment (JamesMBallard) [reply ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.cpp:219]: I agree with @ldanzinger. `setLoading` seems unnecessary when the API itself can tell you this information.

File: ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.h
Comment (JamesMBallard) [inline ArcGISRuntimeSDKQt_CppSamples/Layers/BrowseOGCAPIFeatureService/BrowseOGCAPIFeatureService.h:46]: This can all be inferred from `m_featureLayer` which already has a `loadStatusChanged` signal that notifies. You can simply reuse that logic as needed.

Is you are going to keep this, I would be more specific. Presumably there are a few Loadable objects in play and "loading" is a well-known API term. I would make this `featureLayerLoading` or something like that if you keep it.

**Review by JamesMBallard — COMMENTED** [review]
There should be a simpler way to solve this that stashing bools that mimic the loadStatus of an API object. I would recommend simplifying this if it's possible in the time remaining for 100.12.

**Comment by AndrewBladon** [comment]
Thanks for your suggestions @lsmallwood, @ldanzinger, and @JamesMBallard!

**Comment by AndrewBladon** [comment]
@lsmallwood and @ldanzinger, thanks again for your review comments. We have made the recommended changes. Would it be possible to have the changes reviewed and approved? Thank you.

**Review by ldanzinger — APPROVED** [review]
looks much cleaner. Good to merge

**Review by tanneryould — APPROVED** [review]

**Review by lsmallwood — APPROVED** [review]
LGTM

**Pull request merged**
